### PR TITLE
Allow GradScaler to be pickled

### PR DESCRIPTION
**Pull request by mcarilli**
Should unblock https://github.com/PyTorchLightning/pytorch-lightning/issues/1782.

### CODE REVIEW

**Comment by williamFalcon** [comment]
thank you!

Generally anything that can't be pickled is a problem for us as we automate the ddp init which uses spawn

**Comment by vincentqb** [comment]
@albanD -- would you be willing to review this PR?

**Comment by albanD** [comment]
Sure I'll take a look tomorrow morning

File: test/test_torch.py [hunk 1]
```diff
@@ -3328,6 +3328,32 @@ def test_pickle_function(self):
         b = pickle.loads(serialized)
         self.assertEqual(a, b)
 
+    def test_pickle_gradscaler(self):
```
Comment (ngimel) [inline test/test_torch.py:3331]: can you add it to some of the device-generic tests rather than _TorchMixin? We are trying to get rid of the latter.

**Review by ngimel — APPROVED** [review]
Move the test please, otherwise lgtm.

File: test/test_cuda.py [hunk 2]
```diff
@@ -2150,8 +2151,12 @@ def _run_scaling_case(self, run, unskipped, skipped, atol=1e-7):
             # so any potential errors with the growth factor handling will be magnified.
             scaler = torch.cuda.amp.GradScaler(init_scale=128., growth_factor=2.0, enabled=enabled, growth_interval=1)
 
-            run(data, mod_control, opt_control, scaler, loss_fn, skip_iter, False)
-            run(data, mod_scaling, opt_scaling, scaler, loss_fn, skip_iter, True)
+            _ = run(data, mod_control, opt_control, scaler, loss_fn, skip_iter, False)
+            ret = run(data, mod_scaling, opt_scaling, scaler, loss_fn, skip_iter, True)
+
+            # Allows run() to optionally return a different scaler instance.
+            print(ret)
```
[the remark this answers is not in the document]
Comment (mcarilli) [reply test/test_cuda.py]: Oops, removed

File: torch/cuda/amp/grad_scaler.py [hunk 3]
```diff
@@ -21,6 +21,20 @@ def get(self, device):
         return retval
 
 
+# Defines default_factory for GradScaler's _per_optimizer_states defaultdict,
+# as well as associated "enum" values.  Prefers defining these at top level because
+# - Lambdas can't be pickled, so we don't want to supply a lambda as the factory.
+# - Defining READY, UNSCALED, STEPPED and _refresh_per_optimizer_state within GradScaler
+#   causes a circular reference, which we'd rather avoid.
+READY = 0
```
Comment (albanD) [inline torch/cuda/amp/grad_scaler.py:29]: Do we want to make this an enum now that we don't support python 2 anymore? You removed the comment mentioning this.

**Review by albanD — APPROVED** [review]
Looks good, same comment on tests as natalia.

File: test/test_torch.py (hunk 1 above)
Comment (mcarilli) [reply test/test_torch.py]: Moved test to TestTorchDeviceType, and made it more devoutly device agnostic (see comments in code)

File: torch/cuda/amp/grad_scaler.py (hunk 3 above)
Comment (mcarilli) [reply torch/cuda/amp/grad_scaler.py]: sure, done in https://github.com/pytorch/pytorch/pull/38296/commits/c870c8319d1655276372dda3751e8760558b5e5b.

**Pull request closed without merging**